runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -406,6 +417,56 @@ private void translateImpulse(
     context.addDataStream(Iterables.getOnlyElement(pTransform.getOutputsMap().values()), source);
   }
 
+  @AutoService(NativeTransforms.IsNativeTransform.class)
+  public static class IsFlinkNativeTransform implements NativeTransforms.IsNativeTransform {
+    @Override
+    public boolean test(RunnerApi.PTransform pTransform) {
+      return STREAMING_IMPULSE_TRANSFORM_URL.equals(PTransformTranslation.urnForTransformOrNull(pTransform));
+    }
+  }
+
+  private void translateStreamingImpulse(
+      String id, RunnerApi.Pipeline pipeline, StreamingTranslationContext context) {
+    RunnerApi.PTransform pTransform = pipeline.getComponents().getTransformsOrThrow(id);
+
+    ObjectMapper objectMapper = new ObjectMapper();
+
+    int intervalMillis;
+    int messageCount;
+    try {
+      JsonNode config = objectMapper.readTree(pTransform.getSpec().getPayload().toByteArray());
+      intervalMillis = config.path("interval_ms").asInt(100);
+      messageCount = config.path("message_count").asInt(0);
+    } catch (IOException e) {
+        throw new RuntimeException("Failed to parse configuration for streaming impulse", e);
+    }
+
+    DataStreamSource<WindowedValue<byte[]>> source =
+        context
+            .getExecutionEnvironment()
+            .addSource(
+                new RichParallelSourceFunction<WindowedValue<byte[]>>() {
+                  private AtomicBoolean cancelled = new AtomicBoolean(false);
+                  private AtomicLong count = new AtomicLong();
+
+                  @Override
+                  public void run(SourceContext<WindowedValue<byte[]>> ctx) throws Exception {
+                    while (!cancelled.get() && (messageCount == 0 || count.getAndIncrement() < messageCount)) {
+                      ctx.collect(WindowedValue.valueInGlobalWindow(new byte[] {}));
+                      Thread.sleep(intervalMillis);
```

Comment:
Does the source have to be checkpointing? If so we should synchronize on the checkpoint lock in `ctx` and release it before sleeping.

Reply:
You could also handle `InterruptedException` here since we typically want to continue processing the source, unless `cancel()` has been called.

Reply:
👍